pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+                }
+            };
+            <CurrentActivityTrackingStatus<T>>::put(new_status.clone());
+            Self::deposit_event(Event::<T>::ActivityTrackingStatusSet { status: new_status });
+            Ok(())
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+            // Process the orchestrator chain block author (if it exists) and activity tracking is enabled
+            if let Some(orchestrator_chain_author) = T::GetSelfChainBlockAuthor::get_block_author()
+            {
+                total_weight.saturating_accrue(T::DbWeight::get().reads(1));
+                if let ActivityTrackingStatus::Enabled { start, end: _ } =
+                    <CurrentActivityTrackingStatus<T>>::get()
+                {
+                    total_weight.saturating_accrue(T::DbWeight::get().reads(1));
+                    if start <= T::CurrentSessionIndex::session_index() {
+                        total_weight
+                            .saturating_accrue(Self::on_author_noted(orchestrator_chain_author));
+                    }
+                }
+            }
+            total_weight
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn process_ended_session() {
+            let current_session_index = T::CurrentSessionIndex::session_index();
+            // Since this function will be executed in the beginning of a session
+            // we can populate the active collators from the session that just ended
+            // before resetting the collators storage for the current session
+            // without affecting the current session's active collators records
+            ActiveCollators::<T>::insert(
+                current_session_index.saturating_sub(1),
+                <ActiveCollatorsForCurrentSession<T>>::get(),
+            );
+
+            <ActiveCollatorsForCurrentSession<T>>::put(BoundedBTreeSet::new());
+
+            // Cleanup active collator info for sessions that are older than the maximum allowed
+            if current_session_index > T::MaxInactiveSessions::get() {
+                <crate::pallet::ActiveCollators<T>>::remove(
+                    current_session_index
+                        .saturating_sub(T::MaxInactiveSessions::get())
+                        .saturating_less_one(),
+                );
+            }
+        }
+        pub fn on_author_noted(author: T::CollatorId) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+            let _ = <ActiveCollatorsForCurrentSession<T>>::try_mutate(
+                |active_collators| -> DispatchResult {
+                    if active_collators
+                        .try_insert(author)
+                        .map_err(|_| Error::<T>::MaxCollatorsPerSessionReached)?
```

Comment:
This error is being ignored by the `let _` above, add some logging here so we at least see the log if it happens. Maybe rewrite `let _ =` as `if let Err(e) = ` and log there

Reply:
Actually this is important, if the storage is full it means that we cannot mark any more collators as active, so there will be some collators that produce blocks and will not be included, and will be considered inactive.

Since this should never happen, how about disabling activity tracking in that case? My point is that if the storage is full it means there is a bug in the pallet, so it should be disabled.

Any other ideas @girazoki @chexware ?

Reply:
yes I think that is a good idea. as you said this should never happen, but in the odd event that it does, it's better to disable it directly yes. 

Add a test for it!

Reply:
Added overflow handling and test for it in https://github.com/moondance-labs/tanssi/pull/922/commits/82458129f6464a042fcc1a93b8eeac7c805d8435